R/utils.R

```diff
@@ -132,7 +132,10 @@ eval_with_cols = function(orig_call, all_cols) {
   })
   if (!is.primitive(fun)) {
     named_call = match.call(fun, orig_call)
-    if ("cols" %in% names(formals(fun)) && !"cols" %in% names(named_call)) {
+    if ("cols" %in% names(formals(fun))) {
+      if ("cols" %in% names(named_call)) {
+        stopf("user should not provide cols argument to %s, when specifying the columns for melt or .SDcols; in this context, non-standard evaluation is used internally to set cols to all data table column names, so please fix by removing cols argument", as.character(fun_uneval))
```

Comment:
I think this is a much more serious breaking change than #5247 -- just glancing at GitHub shows lots of people using `cols=`:

https://github.com/search?q=lang%3AR%20%2Fpatterns%5B(%5D%5B%5E)%5D*cols%5Cs*%3D%2F&type=code

Even though I almost exclusively see it done with `cols=names(x)`, such usage is now an error. We need to remove the `cols=` argument more gradually.

Reply:
ok change to warning?

Reply:
Let's map out a deprecation plan, e.g.

1.16.0: `warning()` if `cols` is used _for a subset of columns_, `message()` if it's used _for the full set of columns_. The difference because the latter case should simply drop `cols=`, the former case can actually get different behavior.
1.17.0: Upgrade to error to provide `cols=`
1.18.0: Remove code related to `cols=`

We can consider an option to allow the old behavior to keep working as well, in which case we probably want to lengthen the deprecation cycle by one more release.

Reply:
another option is that we could continue to support user-provided `cols` and modify `measure()` to support that (would be possible I think if we make measure return character column names instead of integer column indices). However I think it would be better to do this warning/error, and encourage users to specify the right regex/pattern arg to patterns/measure. Thoughts?

Reply:
I don't have a strong prior here, though it does sound workable; do you think that would be useful and worth maintaining?

One thought is that it's sometimes clunky to jam everything into a pattern-based approach, e.g. it's not uncommon to have a vector of column names handy already, in which case just plugging it into `cols=` can be convenient as compared to turning that into a regex (esp if the column names might have special regex characters like `(` `{` `.` etc). That said we haven't seen a request to this end.